Arc CDI TCK related fixes, focused around behavior gated behind strict mode

manovotn opened the pull request:
Related to https://github.com/quarkusio/quarkus/issues/28558

This PR tackles tests that we earlier discover needed some `strictMode` behavior (in that Arc is too permissive compared to spec).
It also makes sure we can determine whether we are in strict compat. mode even in runtime.


I know @mkouba mentioned we could also introduce some abstraction for all parameters going into `Arc.initialize(...)` method. I can take a look at that tomorrow even though I am not wholly sold on doing that just yet because there are only two parameters and I am not aware of any more we need ATM. But like I said, I will still happily do that if requested :)

manovotn committed:
Arc - strict mode is not visible in runtime. OM notifications now don't activate req. context in strict mode and only deliver notifications to beans with enabled contexts

manovotn committed:
Arc - conditionally apply strict compatibility mode to bean discovery process

mkouba commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:
```diff
@@ -339,6 +341,12 @@ private void notifyObservers(T event, ObserverExceptionHandler exceptionHandler,
                 Predicate<ObserverMethod<? super T>> predicate) {
             EventContext eventContext = new EventContextImpl<>(event, eventMetadata);
             for (ObserverMethod<? super T> observerMethod : observerMethods) {
+                Bean<?> declaringBean = observerMethod.getDeclaringBean();
```
Is this really necessary? I believe that we handle `Reception.IF_EXISTS` in the generated `ObserverMethod#notify()`...

manovotn replied:
> Is this really necessary? I believe that we handle `Reception.IF_EXISTS` in the generated `ObserverMethod#notify()`...

It is, there are failures otherwise; note that this code is not here for reception handling.
Rather it is there to handle cases where an observer belongs to a bean with context that's inactive - so for instance req. context bean with observer but req. context wasn't activated. Originally, Arc was still trying to invoke it ending with `ContextNotActiveException`.
The dependent and `null` cases are for dependent beans which have no context of their own and for synth. observers which can belong to no bean.

mkouba replied:
I see and I wonder if the spec acutally covers this use case. In any case, I would rather move the logic to the generated `ObserverMethod#notify()` so that it's all in one place.

manovotn replied:
> I see and I wonder if the spec acutally covers this use case

One of the TCKs I uncommented does exactly that - deactivate context, send notification and assert that it did not arrive.

> I would rather move the logic to the generated ObserverMethod#notify() so that it's all in one place.

Hm, to me it makes more sense outside of the method since this is effectively a condition based on which you don't want to notify an observer...

mkouba replied:
> > I see and I wonder if the spec acutally covers this use case
> 
> One of the TCKs I uncommented does exactly that - deactivate context, send notification and assert that it did not arrive.

Oook, it probably makes sense. Is it worth a clarification issue?

> 
> > I would rather move the logic to the generated ObserverMethod#notify() so that it's all in one place.
> 
> Hm, to me it makes more sense outside of the method since this is effectively a condition based on which you don't want to notify an observer...

Logically yes, but from maintainability POV it's better to change this [line](https://github.com/quarkusio/quarkus/blob/main/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverGenerator.java#L417)  to `notify.ifNull(context).trueBranch().returnVoid();` :shrug:

manovotn replied:
This is already in [spec](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#observer_notification):

> If there is no context active for the scope to which the bean declaring the observer method belongs, then the observer method should not be called.

> Logically yes, but from maintainability POV it's better...

Fair enough.

mkouba commented on independent-projects/arc/tcks/cdi-tck-porting-pkg/src/main/java/io/quarkus/arc/tck/porting/ContextsImpl.java:
```diff
@@ -14,13 +14,18 @@
 
 public class ContextsImpl implements Contexts<Context> {
+
+    private InjectableContext.ContextState savedState;
```
Is it guaranteed that `setInactive()`/`setActive()` is called from the same thread? I would at least make the field `volatile`.

manovotn replied:
There is only a few usages in TCK (about 8?) some of which are CDI Full too and in none of them I saw multiple threads being used. But we can surely roll with `volatile` as well.
This is mostly a workaround for what I think should be a TCK challenge I have yet to create.

manovotn replied:
It is very crude indeed :-D
Since I aimed to raise a TCK challenge to begin with, I just went for simplest possible fix. Not to mention, for something like CDI Lite, you can only really toggle on/off req. context and the class already work with `ManagedContext` which will only work for req. context in Arc.
Nonetheless, you are right to point that out; let's do the map approach instead.

mkouba reviewed (commented):
>  I can take a look at that tomorrow even though I am not wholly sold on doing that just yet because there are only two parameters and I am not aware of any more we need ATM. But like I said, I will still happily do that if requested :)

Please do. There will be three `initialize()` methods... too many IMO and I'd hate to add more. I think that we should only keep one and deprecate the others.

manovotn commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:
```diff
@@ -315,8 +317,8 @@ void notify(T event, ObserverExceptionHandler exceptionHandler, boolean async) {
-                // Non-tx observers notifications
-                if (activateRequestContext) {
+                // Non-tx observers notifications; only activate req. context in non-strict mode
+                if (activateRequestContext && !Arc.container().strictCompatibility()) {
```
> You could just modify the `EventImpl#createNotifier()` instead:

Not sure I understand?
The `boolean` present in `createNotifier()` now is used by init a shutdown events where we do not activate the context. Changing it the way you suggest would mean we'd start activating it for those events as well.

mkouba replied:
Yes, I'm talking about the `createNotifier()` without the `activateRequestContext` param which is used for "normal" notifiers and has currently always `activateRequestContext=true`...

Ladicek commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:
```diff
@@ -892,7 +892,7 @@ static ScopeInfo getValidScope(Set<ScopeInfo> scopes, AnnotationTarget target) {
 
     private List<BeanInfo> findBeans(Collection<DotName> beanDefiningAnnotations, List<ObserverInfo> observers,
-            List<InjectionPointInfo> injectionPoints, boolean jtaCapabilities) {
+            List<InjectionPointInfo> injectionPoints, boolean jtaCapabilities, boolean strictCompatibility) {
```
The extra parameter is not necessary.

Ladicek commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java:
```diff
@@ -233,4 +233,12 @@ public interface ArcContainer {
+     *
+     * @return true is strict mode is enabled, false otherwise.
+     */
+    boolean strictCompatibility();
```
Do we want to make this a part of public API?

mkouba replied:
The flag itself is part of the API since we use it to initialize the container. So I think that it's reasonable to add a read method as well.

manovotn replied:
I originally didn't have it exposed but ended up doing that basically with the same reasoning Martin has here.
Not to mention it's nice not having to cast the result of `Arc.container()` to the impl class :)

Ladicek replied:
There's `ArcContainerImpl.instance()` to get the internal type, so that argument doesn't count IMHO :-) But if we need to expose that as a parameter to `Arc.initialize()`, then that makes sense. (For the record, I don't think we do, we could make it part of `ComponentsProvider` or something like that, so that we actually always propagate the build-time status to runtime instead of relying on the integrator to do it correctly. Of course that's a bit more complex.)

manovotn replied:
> (For the record, I don't think we do, we could make it part of ComponentsProvider or something like that, so that we actually always propagate the build-time status to runtime instead of relying on the integrator to do it correctly. Of course that's a bit more complex.)

So this is actually what I started looking into but after chat with Martin, I landed on current solution.

Ladicek commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:
```diff
@@ -339,6 +341,12 @@ private void notifyObservers(T event, ObserverExceptionHandler exceptionHandler,
+                Bean<?> declaringBean = observerMethod.getDeclaringBean();
+                // don't notify observers on beans with inactive context
+                if (declaringBean != null && !declaringBean.getScope().equals(Dependent.class)
+                        && Arc.container().getActiveContext(declaringBean.getScope()) == null) {
```
I have a commit somewhere that adds a context object for the `@Dependent` pseudo-scope, which is always active. We'll be able to remove `!declaringBean.getScope().equals(Dependent.class)` when that is merged, if I understand this correctly.

manovotn replied:
Yes, that is correct - except I will probably move this logic into the generated notify code.
I can add a comment about that.

mkouba replied:
> I have a commit somewhere that adds a context object for the @Dependent pseudo-scope, which is always active.

Hmm, I can't say I'm a fan... why is this needed? `@Dependent` is not a context anyway.

manovotn replied:
Actually, @Ladicek you won't need to remove anything, moving this logic into observer generator doesn't require special case handling as dependent beans always have new instances attached anyway.

manovotn committed:
Arc - incorporate requested changes from PR

manovotn committed:
Arc - introduce ArcInitConfig object along with a builder, replace existing usages, deprecate former init method

manovotn commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Arc.java:
```diff
@@ -12,23 +12,37 @@ public final class Arc {
+     *
+     * @return {@link ArcContainer} instance with default configuration
+     */
     public static ArcContainer initialize() {
```
I kept the no-args variant without deprecation as I think it's better than always requiring config object and passing in the default instance.

manovotn commented:
Ok, round two @mkouba @Ladicek :)

Added two commits, first one addressed comments from this PR, second one introduces `ArcInitConfig` object along with a builder and replaces existing usages.

manovotn commented:
CI failure is unrelated

mkouba commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Arc.java:
```diff
@@ -12,23 +12,37 @@ public final class Arc {
      * @param currentContextFactory
      * @return the initialized container
      */
+    @Deprecated(since = "3.0")
```
Pls add `forRemoval=true`...

mkouba commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:
```diff
@@ -166,10 +166,11 @@ private Notifier<? super T> createNotifier(Class<?> runtimeType) {
     }
 
-    static <T> Notifier<T> createNotifier(Class<?> runtimeType, Type eventType, Set<Annotation> qualifiers,
+    static <T> Notifier<T> createNotifier(Class<?> runtimeType, Type eventType, Set<Annotation> qualifiers, // TODO is this invoked somewhere
```
We don't want to add more `TODO`s ;-). It's a package-private method so it shouldn't be hard to find out...

manovotn replied:
Ah, remnant of a note to myself :D
Removed

Ladicek commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:
```diff
@@ -1026,22 +1039,30 @@ private List<BeanInfo> findBeans(Collection<DotName> beanDefiningAnnotations, Li
                         syncObserverMethods.computeIfAbsent(method, ignored -> new HashSet<>())
                                 .add(beanClass);
                         if (!Modifier.isAbstract(beanClass.flags())) {
-                            // add only concrete classes
```
Nitpicking: this comment clearly refers to the previous line (checking if the class is abstract or "concrete"). The line is moved pretty far away and it's no longer exactly clear what "add only concrete classes" means.

Ladicek commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:
```diff
@@ -1026,22 +1039,30 @@ private List<BeanInfo> findBeans(Collection<DotName> beanDefiningAnnotations, Li
                         asyncObserverMethods.computeIfAbsent(method, ignored -> new HashSet<>())
                                 .add(beanClass);
                         if (!Modifier.isAbstract(beanClass.flags())) {
-                            // add only concrete classes
```
Same here.

Ladicek commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:
```diff
@@ -982,12 +988,19 @@ private List<BeanInfo> findBeans(Collection<DotName> beanDefiningAnnotations, Li
+                            LOGGER.debugf("Producer method found but %s has no bean defining annotation - using @Dependent",
+                                    beanClass);
+                            beanClasses.add(beanClass);
+                        }
```
Shouldn't we do a similar change for producer fields?

mkouba replied:
Good catch!

manovotn replied:
This is now done for producer fields as well

Ladicek commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverGenerator.java:
```diff
@@ -414,8 +413,8 @@ protected void implementNotify(ObserverInfo observer, ClassCreator observerCreat
-                notify.ifNull(context).trueBranch().throwException(ContextNotActiveException.class,
-                        "Context not active: " + observer.getDeclaringBean().getScope().getDotName());
+                // if the context isn't active, don't notify the observer
+                notify.ifNull(context).trueBranch().returnVoid();
```
I'm probably just being stupid here, but conditional observers (`Reception.IF_EXISTS`) are handled above (in the exact same way), so this case only applies to "regular" obsevers (`Reception.ALWAYS`). Why do we need this change? Not saying we don't, just want to understand.

Ladicek replied:
Okay, I see that this probably fixes `ObserverNotificationTest.testObserverMethodNotInvokedIfNoActiveContext`.

Ladicek reviewed (approved):
Few minor comments and one bigger (we should probably treat producer fields similarly to producer methods).

I still think it would be better to pass build-time config to runtime through generated bytecode instead of relying on the integrator to setup the container correctly, but I agree that generating bytecode is generally a pain. And integrators should be sensible :-)

mkouba commented on independent-projects/arc/tcks/cdi-tck-porting-pkg/src/main/java/io/quarkus/arc/tck/porting/ContextsImpl.java:
```diff
@@ -14,13 +16,20 @@
 public class ContextsImpl implements Contexts<Context> {
+
+    // volatile is just a precaution in case some (future) TCK test attempts to use this in between multiple threads
+    private volatile Map<Context, InjectableContext.ContextState> contextStateMap = new HashMap();
```
I agree with both comments.

mkouba replied:
Although it's still a fragile implementation of underspecified SPI... ;-)

manovotn commented:
> Ah and sorry for late review, I need to attend to too many things these days :-)

No problem, I can adjust this PR but prob on Tue due to my PTO and public holidays.
Let me remove the waiting for CI label.

Ladicek reviewed (changes requested):
Changing my review status to _request changes_ to signify that we don't want to merge this yet.

manovotn committed:
Arc - additional requested minor changes

manovotn commented:
@Ladicek updated with requested changes (and rebased onto main), please re-review

Ladicek reviewed (approved):
LGTM, thanks!
